### ISSUE TITLE
crash in triton!triton::engines::symbolic::SymbolicExpression::getFormattedId

### ISSUE THREAD

**Issue by 0vercl0k** [JonathanSalwan/Triton#656]
Hey guys,

The below python script makes the python runtime crashes / or exits on Windows with the latest artefacts:

```python
from triton import *

def sym_exec_gadget(gadget):
    ctx = TritonContext()
    ctx.setArchitecture(ARCH.X86_64)
    ctx.enableMode(MODE.ALIGNED_MEMORY, True)
 
    r = ctx.registers
    gprs = (
        r.rax, r.rbx, r.rcx, r.rdx,
        r.rsi, r.rdi, r.esp, r.ebp,
        r.r8, r.r9, r.r10, r.r11,
        r.r12, r.r13, r.r14
    )

    for gpr in gprs:
        ctx.convertRegisterToSymbolicVariable(gpr)

    for assembly in gadget:
        i = Instruction()
        i.setOpcode(assembly)
        ctx.processing(i)

    srax = ctx.getSymbolicExpressionFromId(
        ctx.getSymbolicRegisterId(
            ctx.registers.rax
        )
    )

    return srax

print sym_exec_gadget(('\x89\xd8', '\xc2\x04\x00'))
```

Here is one of the call-stack I got:

```text
0:000> kc
 # Call Site
00 ntdll!NtTerminateProcess
01 ntdll!RtlExitUserProcess
02 KERNEL32!ExitProcessImplementation
03 MSVCR90!doexit
04 MSVCR90!abort
05 MSVCR90!terminate
06 python!__CxxUnhandledExceptionFilter
07 KERNELBASE!UnhandledExceptionFilter
08 ntdll!RtlUserThreadStart$filt$0
09 ntdll!_C_specific_handler
0a ntdll!RtlpExecuteHandlerForException
0b ntdll!RtlDispatchException
0c ntdll!RtlRaiseException
0d KERNELBASE!RaiseException
0e VCRUNTIME140!_CxxThrowException
0f triton!triton::engines::symbolic::SymbolicExpression::getFormattedId
10 triton!triton::engines::symbolic::operator<<
11 triton!triton::engines::symbolic::operator<<
12 triton!inittriton
13 python27!internal_print
14 python27!PyFile_WriteObject
15 python27!PyEval_EvalFrameEx
16 python27!PyEval_EvalCodeEx
17 python27!run_mod
18 python27!PyRun_FileExFlags
19 python27!PyRun_SimpleFileExFlags
1a python27!PyRun_AnyFileExFlags
1b python27!Py_Main
1c python!__tmainCRTStartup
1d KERNEL32!BaseThreadInitThunk
1e ntdll!RtlUserThreadStart
```

Cheers,
Axel

**Comment by JonathanSalwan** [comment]
Hey axel,

Thanks for this feedback.

> with the latest artefacts:

What artefact branch did you pick? The last one on master is [0.6.2600](https://ci.appveyor.com/project/JonathanSalwan/triton/build/0.6.2600).

**Comment by 0vercl0k** [comment]
I just re-downloaded the artifacts for *a9d4b9a8 *and can confirm this that
the previous file crashes the interpreter on Windows.

Cheers

2018-02-12 1:23 GMT-08:00 Jonathan Salwan <notifications@github.com>:

**Comment by JonathanSalwan** [comment]
Mmmmh. I got it. The problem is that when you print a Triton's object you need a valid `TritonContext`. As you created a `TritonContext` in a function, at the return this context is destroyed (python gc) and then you print a Triton's object (`SymbolicExpression`) where its context is destroyed.

To be honest, I don't know yet how to resolve this kind of behavior.

**Comment by pbrunet** [comment]
It will be fix with the current memory management branch. But the current workaround is certainly to keep the Context alive.

**Comment by 0vercl0k** [comment]
Yeah I figured that that's what was happening :-). Noob question though,
doesn't it mean one of the python object is not ref-counted properly? I
haven't looked at the bindings code yet to be honest so this might very
well be a dumb question :-).

Cheers

2018-02-13 3:36 GMT-08:00 pbrunet <notifications@github.com>:

**Comment by illera88** [comment]
Was this fixed with the used of shared pointers?

**Comment by JonathanSalwan** [comment]
@illera88 it's not the same `shared_ptr`. Currently only `AbstractNode` and `SymbolicExpression` are shared. To solve this issue we have to make `AstContext` or `AstRepresentation` shared too.

**Comment by JonathanSalwan** [comment]
merged into `master` and `dev-v0.7`.

**Comment by JonathanSalwan** [comment]
reopen with 7beb26f7904c95df47e51f4212d151610dead9a0
